File: tests/test_semantic_ui.py
# -*- coding: utf-8 -*-
import pytest
import json
from unittest.mock import AsyncMock, MagicMock
from tests.live_diagnostics import run_semantic_ui_audit
from ui.registry import UI_METADATA

@pytest.mark.asyncio
async def test_semantic_audit_logic():
    """Test semantic audit logic."""
    # Setup mock backend
    mock_backend = MagicMock()
    
    # Mock stream generator
    async def mock_stream(prompt):
        yield "The UI Registry looks "
        yield "well-structured and logically sound. "
        yield "All current actions are appropriate for a local AI assistant."
    
    # Assign the generator to the send_message_async call
    mock_backend.send_message_async.side_effect = lambda p: mock_stream(p)
    
    # Run the audit
    passed = await run_semantic_ui_audit(mock_backend)
    
    # Verify
    assert passed is True
    assert mock_backend.send_message_async.called
    
    # Test a failure case (LLM returns 'error')
    async def mock_error_stream(prompt):
        yield "There is an error in the logic "
        yield "of the settings button mapping."
    
    mock_backend.send_message_async.side_effect = lambda p: mock_error_stream(p)
    passed_with_error = await run_semantic_ui_audit(mock_backend)
    assert passed_with_error is False

@pytest.mark.asyncio
async def test_ui_registry_completeness():
    """Ensure every ID in registry has metadata (Spec #8 Compliance)."""
    from ui.registry import UI_IDS, UI_METADATA
    
    # Get all static attributes of UI_IDS that start with BTN, SW, EXP, or SET
    raw_ids = [getattr(UI_IDS, attr) for attr in dir(UI_IDS) 
               if not attr.startswith('__') and isinstance(getattr(UI_IDS, attr), str)]
    
    for rid in raw_ids:
        assert rid in UI_METADATA, f"UI ID {rid} is missing metadata. Spec violation!"

File: rust_output/src/test_semantic_ui.rs
use anyhow::{Result, Context};
use crate::live_diagnostics::{run_semantic_ui_audit};
use crate::registry::{UI_METADATA};
use tokio;

/// Test semantic audit logic.
pub async fn test_semantic_audit_logic() -> () {
    // Test semantic audit logic.
    let mut mock_backend = MagicMock();
    let mock_stream = |prompt| {
        /* yield "The UI Registry looks ".to_string() */;
        /* yield "well-structured and logically sound. ".to_string() */;
        /* yield "All current actions are appropriate for a local AI assistant.".to_string() */;
    };
    mock_backend::send_message_async.side_effect = |p| mock_stream(p);
    let mut passed = run_semantic_ui_audit(mock_backend).await;
    assert!(passed == true);
    assert!(mock_backend::send_message_async.called);
    let mock_error_stream = |prompt| {
        /* yield "There is an error in the logic ".to_string() */;
        /* yield "of the settings button mapping.".to_string() */;
    };
    mock_backend::send_message_async.side_effect = |p| mock_error_stream(p);
    let mut passed_with_error = run_semantic_ui_audit(mock_backend).await;
    assert!(passed_with_error == false);
}

/// Ensure every ID in registry has metadata (Spec #8 Compliance).
pub async fn test_ui_registry_completeness() -> () {
    // Ensure every ID in registry has metadata (Spec #8 Compliance).
    // TODO: from ui.registry import UI_IDS, UI_METADATA
    let mut raw_ids = dir(UI_IDS).iter().filter(|attr| (!attr.starts_with(&*"__".to_string()) && /* /* isinstance(/* getattr(UI_IDS, attr) */ */ Default::default(), str) */ true)).map(|attr| /* getattr(UI_IDS, attr) */ Default::default()).collect::<Vec<_>>();
    for rid in raw_ids.iter() {
        assert!(UI_METADATA.contains(&rid), "UI ID {} is missing metadata. Spec violation!", rid);
    }
}
